pkg/distributor/forwarding/forwarding.go

```diff
@@ -96,185 +76,286 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
 			Name:      "distributor_forward_samples_total",
 			Help:      "The total number of samples the Distributor forwarded.",
 		}),
+		exemplarsTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Namespace: "cortex",
+			Name:      "distributor_forward_exemplars_total",
+			Help:      "The total number of exemplars the Distributor forwarded.",
+		}),
 		requestLatencyHistogram: promauto.With(reg).NewHistogram(prometheus.HistogramOpts{
 			Namespace: "cortex",
 			Name:      "distributor_forward_requests_latency_seconds",
 			Help:      "The client-side latency of requests to forward metrics made by the Distributor.",
 			Buckets:   []float64{.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10, 20, 30},
 		}),
 	}
-}
 
-func (r *forwarder) NewRequest(ctx context.Context, tenant string, rules validation.ForwardingRules) Request {
-	return &request{
-		ctx:    ctx,
-		client: &r.client, // http client should be re-used so open connections get re-used.
-		pools:  &r.pools,
+	f.Service = services.NewBasicService(f.start, f.loop, f.stop)
 
-		tsByEndpoint: make(map[string]*[]mimirpb.PreallocTimeseries),
+	return f
+}
 
-		rules:           rules,
-		timeout:         r.cfg.RequestTimeout,
-		propagateErrors: r.cfg.PropagateErrors,
+func (f *forwarder) start(ctx context.Context) error {
+	f.workerWg.Add(f.cfg.RequestConcurrency)
 
-		requests: r.requestsTotal,
-		errors:   r.errorsTotal,
-		samples:  r.samplesTotal,
-		latency:  r.requestLatencyHistogram,
+	for i := 0; i < f.cfg.RequestConcurrency; i++ {
+		go f.worker()
 	}
+
+	return nil
 }
 
-type request struct {
-	ctx    context.Context
-	client *http.Client
-	pools  *pools
+func (f *forwarder) loop(ctx context.Context) error {
```

Comment:
nit: If the only thing the "loop" method does you can `services.NewIdleService(start, stop)`

Reply:
thanks! fixed c4b39b0c7e978d6a50e75a79409d82104ff1510e